File: dbk/dbk/report/warehouse_stock_out_report/warehouse_stock_out_report.py
# Copyright (c) 2025, DBK and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt, getdate, nowdate

def execute(filters=None):
	columns = get_columns()
	data = get_data(filters)
	return columns, data

def get_columns():
	"""Define report columns"""
	return [
		
		{
			"fieldname": "date",
			"label": _("Date"),
			"fieldtype": "Date",
			"width": 100
		},
		{
			"fieldname": "department",
			"label": _("Item Group"),
			"fieldtype": "Data",
			"width": 150
		},
		{
			"fieldname": "item_code",
			"label": _("Item"),
			"fieldtype": "Link",
			"options": "Item",
			"width": 120
		},
		{
			"fieldname": "quantity",
			"label": _("Quantity"),
			"fieldtype": "Float",
			"width": 100
		},
		{
			"fieldname": "uom",
			"label": _("Unit"),
			"fieldtype": "Link",
			"options": "UOM",
			"width": 80
		},
		{
			"fieldname": "from_warehouse",
			"label": _("From Warehouse"),
			"fieldtype": "Link",
			"options": "Warehouse",
			"width": 180
		},
		{
			"fieldname": "to_warehouse",
			"label": _("Transferred To"),
			"fieldtype": "Link",
			"options": "Warehouse",
			"width": 180
		},
		{
			"fieldname": "comment",
			"label": _("Comment"),
			"fieldtype": "Small Text",
			"width": 200
		},
		{
			"fieldname": "purpose",
			"label": _("Purpose"),
			"fieldtype": "Data",
			"width": 120
		},
		{
			"fieldname": "voucher_no",
			"label": _("Stock Entry"),
			"fieldtype": "Link",
			"options": "Stock Entry",
			"width": 140
		},
		{
			"fieldname": "material_request",
			"label": _("Material Request"),
			"fieldtype": "Link",
			"options": "Material Request",
			"width": 140
		}
	]

def get_data(filters):
	"""Get stock out data based on filters"""
	conditions = get_conditions(filters)
	
	# Check if warehouse is a group and get child warehouses
	warehouses = get_warehouses(filters.get("warehouse"))
	
	if not warehouses:
		return []
	
	warehouse_conditions = "AND sed.s_warehouse IN ({})".format(
		", ".join([f"'{w}'" for w in warehouses])
	)
	
	query = f"""
		SELECT
			COALESCE(item.item_group, 'Uncategorized') as department,
			sed.item_code,
			sed.uom,
			sed.qty as quantity,
			se.posting_date as date,
			sed.s_warehouse as from_warehouse,
			sed.t_warehouse as to_warehouse,
			se.stock_entry_type as purpose,
			mr.custom_material_usage_intention as comment,
			mr.name as material_request,
			se.name as voucher_no,
			se.material_request
		FROM
			`tabStock Entry Detail` sed
		INNER JOIN
			`tabStock Entry` se ON sed.parent = se.name
		INNER JOIN
			`tabItem` item ON sed.item_code = item.item_code
		LEFT JOIN
			`tabMaterial Request` mr ON se.material_request = mr.name
		WHERE
			se.docstatus = 1
			AND sed.s_warehouse IS NOT NULL
			AND sed.s_warehouse != ''
			AND se.stock_entry_type IN ('Material Transfer', 'Material Issue')
			{warehouse_conditions}
			{conditions}
		ORDER BY
			se.posting_date DESC, se.posting_time DESC
	"""
	
	data = frappe.db.sql(query, as_dict=1)
	
	# If warehouse is a group, aggregate data
	if filters.get("warehouse") and filters.get("aggregate_group"):
		data = aggregate_warehouse_data(data, filters)
	
	return data

def get_conditions(filters):
	"""Build SQL conditions based on filters"""
	conditions = []
	
	if filters.get("from_date"):
		conditions.append(f"se.posting_date >= '{filters.get('from_date')}'")
	
	if filters.get("to_date"):
		conditions.append(f"se.posting_date <= '{filters.get('to_date')}'")
	
	if filters.get("department"):
		conditions.append(f"item.item_group = '{filters.get('department')}'")
	
	if filters.get("item_code"):
		conditions.append(f"sed.item_code = '{filters.get('item_code')}'")
	
	if filters.get("item_group"):
		conditions.append(f"item.item_group = '{filters.get('item_group')}'")
	
	if filters.get("purpose"):
		conditions.append(f"se.stock_entry_type = '{filters.get('purpose')}'")
	
	if filters.get("material_request"):
		conditions.append(f"se.material_request = '{filters.get('material_request')}'")
	
	return "AND " + " AND ".join(conditions) if conditions else ""

def get_warehouses(warehouse):
	"""Get warehouse and its children if it's a group"""
	if not warehouse:
		# Return all warehouses if none selected
		return [w.name for w in frappe.get_all("Warehouse", filters={"is_group": 0})]
	
	# Check if warehouse is a group
	is_group = frappe.db.get_value("Warehouse", warehouse, "is_group")
	
	if is_group:
		# Get all child warehouses
		lft, rgt = frappe.db.get_value("Warehouse", warehouse, ["lft", "rgt"])
		child_warehouses = frappe.db.sql("""
			SELECT name 
			FROM `tabWarehouse`
			WHERE lft >= %s AND rgt <= %s AND is_group = 0
		""", (lft, rgt), as_dict=1)
		return [w.name for w in child_warehouses]
	else:
		return [warehouse]

def aggregate_warehouse_data(data, filters):
	"""Aggregate data by item when warehouse is a group"""
	if not filters.get("aggregate_group"):
		return data
	
	aggregated = {}
	
	for row in data:
		key = (row.item_code, row.department, row.date, )
		
		if key in aggregated:
			aggregated[key]["quantity"] += row.quantity
			# Concatenate warehouses if different
			if row.from_warehouse and row.from_warehouse not in aggregated[key]["from_warehouse"]:
				aggregated[key]["from_warehouse"] += f", {row.from_warehouse}"
			if row.to_warehouse and row.to_warehouse not in aggregated[key]["to_warehouse"]:
				aggregated[key]["to_warehouse"] += f", {row.to_warehouse}"
			# Concatenate voucher numbers
			if row.voucher_no and row.voucher_no not in aggregated[key]["voucher_no"]:
				aggregated[key]["voucher_no"] += f", {row.voucher_no}"
		else:
			aggregated[key] = row.copy()
	
	return list(aggregated.values())

def get_chart_data(data, filters):
	"""Generate chart data for the report"""
	if not data:
		return None
	
	# Group by department
	department_data = {}
	for row in data:
		dept = row.get("department") or "Uncategorized"
		if dept not in department_data:
			department_data[dept] = 0
		department_data[dept] += flt(row.get("quantity", 0))
	
	labels = list(department_data.keys())
	values = list(department_data.values())
	
	return {
		"data": {
			"labels": labels,
			"datasets": [
				{
					"name": "Quantity Issued",
					"values": values
				}
			]
		},
		"type": "bar",
		"colors": ["#4C51BF"]
	}

File: dbk/dbk/report/warehouse_stock_out_report/warehouse_stock_out_report.js
// Copyright (c) 2025, DBK and contributors
// For license information, please see license.txt

frappe.query_reports["Warehouse Stock Out Report"] = {
	"filters": [
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
			"reqd": 1
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"reqd": 1
		},
		{
			"fieldname": "warehouse",
			"label": __("Warehouse"),
			"fieldtype": "Link",
			"options": "Warehouse",
			"get_query": function() {
				return {
					"filters": {}
				};
			}
		},
		{
			"fieldname": "aggregate_group",
			"label": __("Aggregate Group Warehouses"),
			"fieldtype": "Check",
			"default": 1,
			"description": "If warehouse is a group, aggregate data from all child warehouses"
		},
		{
			"fieldname": "department",
			"label": __("Department (Item Group)"),
			"fieldtype": "Link",
			"options": "Item Group"
		},
		{
			"fieldname": "item_code",
			"label": __("Item"),
			"fieldtype": "Link",
			"options": "Item",
			"get_query": function() {
				return {
					"query": "frappe.desk.search.search_link",
					"filters": {}
				};
			}
		},
		{
			"fieldname": "item_group",
			"label": __("Item Group"),
			"fieldtype": "Link",
			"options": "Item Group"
		},
		{
			"fieldname": "purpose",
			"label": __("Purpose"),
			"fieldtype": "Select",
			"options": "\nMaterial Transfer\nMaterial Issue",
			"default": ""
		},
		{
			"fieldname": "material_request",
			"label": __("Material Request"),
			"fieldtype": "Link",
			"options": "Material Request"
		}
	],
	
	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		
		// Highlight quantities
		if (column.fieldname == "quantity" && data) {
			value = `<span style="color: #e53e3e; font-weight: 600;">${value}</span>`;
		}
		
		// Format department with badge
		if (column.fieldname == "department" && data && data.department) {
			const colors = {
				'MVM': 'badge-primary',
				'GENERAL': 'badge-success',
				'Tailoring': 'badge-info',
				'Carpentry': 'badge-warning',
				'Uncategorized': 'badge-secondary'
			};
			const badge_class = colors[data.department] || 'badge-primary';
			value = `<span class="badge ${badge_class}">${data.department}</span>`;
		}
		
		// Format warehouse names with shorter display
		if ((column.fieldname == "from_warehouse" || column.fieldname == "to_warehouse") && value) {
			// Show shortened warehouse name if too long
			const display_value = value.length > 20 ? value.substring(0, 20) + '...' : value;
			value = `<span style="font-size: 0.85em;" title="${value}">${display_value}</span>`;
		}
		
		// Format comment with tooltip for long text
		if (column.fieldname == "comment" && value) {
			const display_value = value.length > 50 ? value.substring(0, 50) + '...' : value;
			value = `<span title="${value}">${display_value}</span>`;
		}
		
		return value;
	},
	
	"onload": function(report) {
		// Add custom buttons
		report.page.add_inner_button(__("Export to Excel"), function() {
			frappe.query_report.export_report("xlsx");
		});
		
		report.page.add_inner_button(__("Print"), function() {
			frappe.query_report.print_report();
		});
		
		// Add summary section
		report.page.add_inner_button(__("Show Summary"), function() {
			show_summary(report);
		});
		
		// Add refresh button
		report.page.add_inner_button(__("Refresh"), function() {
			report.refresh();
		});
	},
	
	"after_datatable_render": function(datatable) {
		// Add summary row at the bottom
		$(datatable.wrapper).find('.dt-scrollable').append(`
			<div class="dt-summary" style="padding: 10px; background: #f7fafc; border-top: 2px solid #e2e8f0; margin-top: 10px;">
				<strong>Total Records: </strong><span id="total-records">0</span> | 
				<strong>Total Quantity: </strong><span id="total-quantity">0</span>
			</div>
		`);
	},
	
	"get_datatable_options": function(options) {
		return Object.assign(options, {
			checkboxColumn: true,
			events: {
				onCheckRow: function(data) {
					// Handle row selection
					console.log(data);
				}
			}
		});
	}
};

function show_summary(report) {
	let data = report.data;
	
	if (!data || data.length === 0) {
		frappe.msgprint(__("No data to summarize"));
		return;
	}
	
	// Calculate summaries
	let total_quantity = 0;
	let department_summary = {};
	let warehouse_summary = {};
	let item_summary = {};
	let purpose_summary = {};
	
	data.forEach(row => {
		total_quantity += row.quantity || 0;
		
		// By department
		if (row.department) {
			department_summary[row.department] = (department_summary[row.department] || 0) + row.quantity;
		}
		
		// By warehouse
		if (row.from_warehouse) {
			warehouse_summary[row.from_warehouse] = (warehouse_summary[row.from_warehouse] || 0) + row.quantity;
		}
		
		// By purpose
		if (row.purpose) {
			purpose_summary[row.purpose] = (purpose_summary[row.purpose] || 0) + row.quantity;
		}
		
		// By item
		if (row.item_code) {
			if (!item_summary[row.item_code]) {
				item_summary[row.item_code] = {
					name: row.item_name || row.item_code,
					quantity: 0
				};
			}
			item_summary[row.item_code].quantity += row.quantity;
		}
	});
	
	// Build summary HTML
	let html = `
		<div style="padding: 20px;">
			<h4 style="margin-bottom: 20px; color: #2d3748;">📊 Stock Out Summary</h4>
			
			<div style="margin-bottom: 25px;">
				<h5 style="color: #4a5568; margin-bottom: 10px;">Overall Statistics</h5>
				<table class="table table-bordered" style="width: 100%;">
					<tr>
						<td style="background: #f7fafc; font-weight: 600; width: 40%;">Total Transactions</td>
						<td>${data.length}</td>
					</tr>
					<tr>
						<td style="background: #f7fafc; font-weight: 600;">Total Quantity Issued</td>
						<td style="color: #e53e3e; font-weight: 700;">${total_quantity.toFixed(2)}</td>
					</tr>
				</table>
			</div>
			
			<div style="margin-bottom: 25px;">
				<h5 style="color: #4a5568; margin-bottom: 10px;">By Purpose</h5>
				<table class="table table-bordered" style="width: 100%;">
					<thead style="background: #edf2f7;">
						<tr>
							<th>Purpose</th>
							<th style="text-align: right;">Quantity</th>
							<th style="text-align: right;">% of Total</th>
						</tr>
					</thead>
					<tbody>
						${Object.entries(purpose_summary)
							.sort((a, b) => b[1] - a[1])
							.map(([purpose, qty]) => `
								<tr>
									<td><span class="badge badge-info">${purpose}</span></td>
									<td style="text-align: right; font-weight: 600;">${qty.toFixed(2)}</td>
									<td style="text-align: right;">${((qty / total_quantity) * 100).toFixed(1)}%</td>
								</tr>
							`).join('')}
					</tbody>
				</table>
			</div>
			
			<div style="margin-bottom: 25px;">
				<h5 style="color: #4a5568; margin-bottom: 10px;">By Department (Item Group)</h5>
				<table class="table table-bordered" style="width: 100%;">
					<thead style="background: #edf2f7;">
						<tr>
							<th>Department</th>
							<th style="text-align: right;">Quantity</th>
							<th style="text-align: right;">% of Total</th>
						</tr>
					</thead>
					<tbody>
						${Object.entries(department_summary)
							.sort((a, b) => b[1] - a[1])
							.map(([dept, qty]) => `
								<tr>
									<td><span class="badge badge-primary">${dept}</span></td>
									<td style="text-align: right; font-weight: 600;">${qty.toFixed(2)}</td>
									<td style="text-align: right;">${((qty / total_quantity) * 100).toFixed(1)}%</td>
								</tr>
							`).join('')}
					</tbody>
				</table>
			</div>
			
			<div style="margin-bottom: 25px;">
				<h5 style="color: #4a5568; margin-bottom: 10px;">Top 10 Items by Quantity</h5>
				<table class="table table-bordered" style="width: 100%;">
					<thead style="background: #edf2f7;">
						<tr>
							<th>Item</th>
							<th style="text-align: right;">Quantity</th>
						</tr>
					</thead>
					<tbody>
						${Object.entries(item_summary)
							.sort((a, b) => b[1].quantity - a[1].quantity)
							.slice(0, 10)
							.map(([code, data]) => `
								<tr>
									<td>${data.name}</td>
									<td style="text-align: right; font-weight: 600;">${data.quantity.toFixed(2)}</td>
								</tr>
							`).join('')}
					</tbody>
				</table>
			</div>
			
			<div>
				<h5 style="color: #4a5568; margin-bottom: 10px;">By Source Warehouse</h5>
				<table class="table table-bordered" style="width: 100%;">
					<thead style="background: #edf2f7;">
						<tr>
							<th>Warehouse</th>
							<th style="text-align: right;">Quantity</th>
							<th style="text-align: right;">% of Total</th>
						</tr>
					</thead>
					<tbody>
						${Object.entries(warehouse_summary)
							.sort((a, b) => b[1] - a[1])
							.map(([warehouse, qty]) => `
								<tr>
									<td>${warehouse}</td>
									<td style="text-align: right; font-weight: 600;">${qty.toFixed(2)}</td>
									<td style="text-align: right;">${((qty / total_quantity) * 100).toFixed(1)}%</td>
								</tr>
							`).join('')}
					</tbody>
				</table>
			</div>
		</div>
	`;
	
	// Show in dialog
	let d = new frappe.ui.Dialog({
		title: __("Stock Out Summary"),
		size: "large",
		fields: [
			{
				fieldtype: "HTML",
				fieldname: "summary_html"
			}
		]
	});
	
	d.fields_dict.summary_html.$wrapper.html(html);
	d.show();
}

// Update summary when data changes
frappe.query_reports["Warehouse Stock Out Report"].on_data_render = function(data) {
	if (data && data.length > 0) {
		let total_quantity = data.reduce((sum, row) => sum + (row.quantity || 0), 0);
		
		setTimeout(() => {
			$('#total-records').text(data.length);
			$('#total-quantity').text(total_quantity.toFixed(2));
		}, 100);
	}
};